Load saved custom Loss class

thierryherrmann opened the pull request:
Attempt to fix https://github.com/tensorflow/tensorflow/issues/32612

Here's an example with a custom Loss class, including config to ensure it's properly reloaded:
```python
class HuberLoss(keras.losses.Loss):
    def __init__(self, threshold=1.0, **kwargs):
        super().__init__(**kwargs)
        self.threshold = threshold
    
    @tf.function
    def call(self, y_true, y_pred):
        error = tf.abs(y_true - y_pred)
        is_small_error = error <= self.threshold
        squared_loss = tf.square(error) / 2
        linear_loss = error * self.threshold - 0.5 * self.threshold**2
        return tf.where(is_small_error, squared_loss, linear_loss)
    
    def get_config(self):
        cfg = super().get_config()
        cfg['threshold'] = self.threshold
        return cfg
```
The class can be used for a regression task this way:
```python
model = keras.Sequential([
  keras.layers.Dense(30, activation='relu', input_shape=X_train.shape[1:]),
  keras.layers.Dense(1)
])

model.compile(loss=HuberLoss(2.0), optimizer="sgd")
model.fit(X_train, y_train, epochs=1, validation_data=(X_valid, y_valid))

# save and reload
model.save('model_with_huber_loss_class.h5')
model = keras.models.load_model('model_with_huber_loss_class.h5', custom_objects={'HuberLoss': HuberLoss})

# continue training
model.fit(X_train, y_train, epochs=1, validation_data=(X_valid, y_valid))
```

The problem was that `saving_util.compile_args_from_training_config(training_config, custom_objects=None)` supports `custom_objects` but `custom_objects` was not propagated to `losses.get()`.
`losses.get()` later called `deserialize(identifier)` without specifying `custom_objects` while the `custom_objects` argument is supported by `deserialize()`: the default `None` was used and the custom class load failed.

thierryherrmann committed:
fix for #32612 : load saved custom Loss class

omalleyt12 commented on tensorflow/python/keras/saving/saving_utils.py:
```diff
@@ -224,7 +224,7 @@ def compile_args_from_training_config(training_config, custom_objects=None):
   loss_config = training_config['loss']  # Deserialize loss class.
   if isinstance(loss_config, dict) and 'class_name' in loss_config:
-    loss_config = losses.get(loss_config)
+    loss_config = losses.get(loss_config, custom_objects)
```
I believe the fix here should be to replace lines 226-234 with:

```python
if isinstance(loss_config, dict) and 'class_name' in loss_config:
  loss = losses.deserialize(loss_config, custom_objects)
else:
  loss = nest.map_structure(
    lambda obj: losses.deserialize(obj, custom_objects),
    loss_config)

metrics_config = training_config['metrics']
if isinstance(metrics_config, dict) and 'class_name' in metrics_config:
  metrics = metrics.deserialize(metrics_config, custom_objects)
else:
  metrics = nest.map_structure(
    lambda obj: metrics.deserialize(obj, custom_objects)
    metrics_config)

weighted_metrics_config = training_config['weighted_metrics']
if isinstance(weighted_metrics_config, dict) and 'class_name' in weighted_metrics_config:
  weighted_metrics = metrics.deserialize(weighted_metrics_config, custom_objects)
else:
  weighted_metrics = nest.map_structure(
    lambda obj: metrics.deserialize(obj, custom_objects),
    weighted_metrics_config)
```

Could you please make these changes and confirm this works for your use case?

thierryherrmann replied:
thanks @omalleyt12 
will do.

omalleyt12 commented on tensorflow/python/keras/losses.py:
```diff
@@ -1176,14 +1176,14 @@ def deserialize(name, custom_objects=None):
 
 @keras_export('keras.losses.get')
-def get(identifier):
+def get(identifier, custom_objects=None):
```
Please see the below comment, I think this change is not needed

omalleyt12 reviewed (changes requested):
Thanks for the PR! Please see the suggested changes below

thierryherrmann committed:
fix for #32612 : apply suggested changes

thierryherrmann committed:
fix for #32612 : apply suggested changes

thierryherrmann commented:
@omalleyt12 
I tested the changes you've suggested (and pushed them to this PR).
I only had to rename `metrics` --> `metrics_obj` to not conflict with the `metrics` module.
It works indeed: `losses.deserialize(...)` is called and
the loss custom object is correctly deserialized. I can resume training with the reloaded model.
```python
  if isinstance(loss_config, dict) and 'class_name' in loss_config:
    loss = losses.deserialize(loss_config, custom_objects)  # <<< this is executed
```


Since your proposed changes also involve metrics, I've tested with a custom metrics class.
I wrote a class HuberMetric(keras.metrics.Metric) for the test.
Full code: https://gist.github.com/thierryherrmann/550906c4f97ff685cf21b0418a485063
Model training goes well, then I save the model but it fails on reload.

In this part of `saving_utils.py`:
```python
  metrics_config = training_config['metrics']
  if isinstance(metrics_config, dict) and 'class_name' in metrics_config:
    metrics_obj = metrics.deserialize(metrics_config, custom_objects)
  else:
    metrics_obj = nest.map_structure(
      lambda obj: metrics.deserialize(obj, custom_objects),
      metrics_config)
```
`metrics_config` is not an instance of dict but of list. Indeed it is this:
`[{'class_name': 'HuberMetric', 'config': {'dtype': 'float32', 'name': 'huber_metric', 'threshold': 0.2}}]`
So the code executes the else part and `nest.map_structure(...)`
and fails when trying to deserialize 'float32'. 
I can provide more details but looks like we're drifting from the original bug that is about reloading custom losses.

In summary (excerpt from the gist above):
```python
if True:
    # LOSS custom class: FIXED by the PR (thanks @omalleyt12)
    model.compile(loss=HuberLoss(2.0), optimizer="sgd")
    model.fit(X_train, y_train, epochs=1, validation_data=(X_valid, y_valid))
    model.save('model.h5')
    model = keras.models.load_model('model.h5', custom_objects={'HuberLoss': HuberLoss})

if False:
    # LOSS custom function: was working before the PR, still works with the PR
    model.compile(loss=huber_fn, optimizer="sgd")
    model.fit(X_train, y_train, epochs=1, validation_data=(X_valid, y_valid))
    model.save('model.h5')
    model = keras.models.load_model('model.h5', custom_objects={'huber_fn': huber_fn})

if False:
    # METRICS custom class: was broken before the PR, still broken with the PR
    model.compile(loss=keras.losses.mean_squared_error, optimizer="sgd", metrics=[HuberMetric(threshold=.2)])
    model.fit(X_train, y_train, epochs=1, validation_data=(X_valid, y_valid))
    model.save('model.h5')
    model = keras.models.load_model('model.h5', custom_objects={'HuberMetric':HuberMetric})

if False:
    # METRICS custom function: was working before the PR, still works with the PR
    model.compile(loss=keras.losses.mean_squared_error, optimizer="sgd", metrics=[huber_fn])
    model.fit(X_train, y_train, epochs=1, validation_data=(X_valid, y_valid))
    model.save('model.h5')
    model = keras.models.load_model('model.h5', custom_objects={'huber_fn':huber_fn})

# continue training with the reloaded model
model.fit(X_train, y_train, epochs=1, validation_data=(X_valid, y_valid))
```

So the PR with your suggested changes fixes the original bug and doesn't break things.
I'll leave up to reviewers if this should be merged. Custom metrics class reloading might be considered as another issue.
Thanks

omalleyt12 reviewed (dismissed):
Thanks for the PR! LGTM!

gbaned commented:
@thierryherrmann Could you please check failed build errors? Thanks!

omalleyt12 commented:
> So the PR with your suggested changes fixes the original bug and doesn't break things.
> I'll leave up to reviewers if this should be merged. Custom metrics class reloading might be considered as another issue.
> Thanks

Hm thanks for looking into this, yes the issue is the `nest.map_structure`. Instead it should be safer to assume that metrics_config is either a dict or a list of dicts. This should probably be the same for weighted_metrics and indeed for losses as well:

```python
metrics_config = training_config['metrics']
  if isinstance(metrics_config, dict) and 'class_name' in metrics_config:
    metrics_obj = metrics.deserialize(metrics_config, custom_objects)
  else:
    metrics_obj = [
      lambda obj: metrics.deserialize(obj, custom_objects)
      for obj in metrics_config]
```

thierryherrmann commented:
@omalleyt12 agreed. I'll look at this over the week-end. Thanks again

thierryherrmann committed:
fix for #32612 : fix unit tests

thierryherrmann commented:
Thanks again for your suggestion @omalleyt12. I just removed the lambda from your proposal and it works. You've done all the work.

I've added `elif isinstance(..., list)` for the `metrics_config` and `weighted_metrics_config` but not for losses since my understanding is that we can only have one loss.

Also I've left an `else` fallback executing the original `nest.map_structure(...)` in case the original designers rely on this behavior now or in the future.
This fallback branch is actually still executed if we specify no metric or weigthted metric (in which case `metrics_config` (resp. `weighted_metrics_config`) are None)

All failing tests pass now. But @omalleyt12, is it just me who has a very hard time to run the test suite with `bazel test` (either by building on my laptop directly or by using a docker image) ? I actually executed the tests by taking a copy of `tensorflow/python/keras/saving/hd5_format_test.py` and execute it separately.

omalleyt12 reviewed (dismissed):
@thierryherrmann Thanks for the PR! Looks good to me!

pavithrasv commented:
Thanks a lot for the changes @thierryherrmann and @omalleyt12 . Sorry i am late in the game, is it possible to add tests for the custom loss and metric configs?

thierryherrmann commented:
@pavithrasv I'm a bit busy at the moment. Should be ok for this week-end. Let me know if you really need the tests earlier.

pavithrasv commented:
Oh ok, no worries then, we can get this submitted and add the tests as soon as possible after that. Thank you!

gbaned commented:
@thierryherrmann Could you please check failed build errors? Thanks!

pavithrasv commented:
@thierryherrmann I took a look at the failures:

1. There is a circular dependency with the import of metrics.
2. There are linter errors: There is a 80 character limit for every line.
3. failure in losses_test
4. failure in integration_test

For the test case `test_serializing_model_with_loss_class`, the loss_config looks like:
`<type 'dict'>: {u'model_output': {u'class_name': u'_MSEMAELoss', u'config': {u'mse_fraction': 0.3}}}`

`if isinstance(loss_config, dict) and 'class_name' in loss_config:` will fail when the loss input to compile is given as a dictionary: For example: {'model_output': my_loss_class(...)}. There are other loss and metric configurations for which we will see failures.

I think the following fix should handle all the above issues, can you please try it? 

```
def convert_output_metrics(metrics_config, custom_objects):
  from google3.third_party.tensorflow.python.keras import metrics as metrics_module  # pylint:disable=g-import-not-at-top
  if isinstance(metrics_config, list):
    return [convert_output_metrics(mc, custom_objects) for mc in metrics_config]
  elif (isinstance(metrics_config, dict) or
        (metrics_config not in ['accuracy', 'acc', 'crossentropy', 'ce'])):
    # Do not deserialize accuracy and cross-entropy strings as we have special
    # case handling for these in compile, based on model output shape.
    return metrics_module.deserialize(metrics_config, custom_objects)
  return metrics_config


def compile_args_from_training_config(training_config, custom_objects=None):
  """Return model.compile arguments from training config."""
  if custom_objects is None:
    custom_objects = {}

  optimizer_config = training_config['optimizer_config']
  optimizer = optimizers.deserialize(
      optimizer_config, custom_objects=custom_objects)

  # Recover losses.
  loss_config = training_config['loss']
  if isinstance(loss_config, list):  # Loss fed to compile as a list.
    loss = [losses.deserialize(lc, custom_objects) for lc in loss_config]
  elif isinstance(loss_config, dict) and 'class_name' not in loss_config:
    # Loss fed to compile as a dict.
    loss = {
        k: losses.deserialize(v, custom_objects)
        for (k, v) in loss_config.items()
    }
  else:  # Loss fed to compile as a str/ function/ class instance.
    loss = losses.deserialize(loss_config, custom_objects)

  # Recover metrics.
  metrics_config = training_config['metrics']
  if isinstance(metrics_config, dict):  # Metrics fed to compile as a dict.
    metrics = {
        k: convert_output_metrics(v, custom_objects)
        for (k, v) in metrics_config.items()
    }
  elif isinstance(metrics_config, list):  # Metrics fed to compile as a list.
    metrics = [
        convert_output_metrics(m, custom_objects) for m in metrics_config
    ]
  else:  # No metrics.
    metrics = None

  # Recover weighted metrics.
  weighted_metrics_config = training_config['weighted_metrics']
  if isinstance(weighted_metrics_config, dict):
    # Metrics fed to compile as a dict.
    weighted_metrics = {
        k: convert_output_metrics(v, custom_objects)
        for (k, v) in weighted_metrics_config.items()
    }
  elif isinstance(weighted_metrics_config, list):
    # Metrics fed to compile as a list.
    weighted_metrics = [
        convert_output_metrics(m, custom_objects)
        for m in weighted_metrics_config
    ]
  else:  # No metrics.
    weighted_metrics = None

  sample_weight_mode = training_config['sample_weight_mode']
  loss_weights = training_config['loss_weights']

  return dict(
      optimizer=optimizer,
      loss=loss,
      metrics=metrics,
      weighted_metrics=weighted_metrics,
      loss_weights=loss_weights,
      sample_weight_mode=sample_weight_mode)
```

thierryherrmann commented:
@pavithrasv yes I will try, thanks. Sorry I didn't find the time so far but I'll try to fix it.

pavithrasv commented:
Small edit to the above fix: 
Please replace `metrics_config = training_config['metrics']` with 
`metrics_config = training_config.get('metrics', None)` and 
`weighted_metrics_config = training_config['weighted_metrics']` with 
`weighted_metrics_config = training_config.get('weighted_metrics', None)`

Thank you!

thierryherrmann commented:
@pavithrasv can you contact me at thierry.herrmann@gmail.com ?

pavithrasv commented:
Sure, left you a message, please feel free to email me if you have any questions. Thank you!

thierryherrmann committed:
fix for #32612 : apply suggested changes

thierryherrmann committed:
fix for #32612 : apply suggested changes

thierryherrmann committed:
fix for #32612 : apply suggested changes

pavithrasv commented on tensorflow/python/keras/saving/saving_utils.py:
```diff
@@ -219,6 +219,18 @@ def should_overwrite(filepath, overwrite):
 
 
+def convert_output_metrics(metrics_config, custom_objects):
+  from google3.third_party.tensorflow.python.keras import metrics as metrics_module  # pylint:disable=g-import-not-at-top
```
This can be removed.
from tensorflow.python.keras import metrics as metrics_module  # pylint:disable=g-import-not-at-top

thierryherrmann committed:
fix for #32612 : apply suggested changes
